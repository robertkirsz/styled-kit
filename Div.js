module.exports=function(e){var t={};function n(r){if(t[r])return t[r].exports;var i=t[r]={i:r,l:!1,exports:{}};return e[r].call(i.exports,i,i.exports,n),i.l=!0,i.exports}return n.m=e,n.c=t,n.d=function(e,t,r){n.o(e,t)||Object.defineProperty(e,t,{enumerable:!0,get:r})},n.r=function(e){"undefined"!=typeof Symbol&&Symbol.toStringTag&&Object.defineProperty(e,Symbol.toStringTag,{value:"Module"}),Object.defineProperty(e,"__esModule",{value:!0})},n.t=function(e,t){if(1&t&&(e=n(e)),8&t)return e;if(4&t&&"object"==typeof e&&e&&e.__esModule)return e;var r=Object.create(null);if(n.r(r),Object.defineProperty(r,"default",{enumerable:!0,value:e}),2&t&&"string"!=typeof e)for(var i in e)n.d(r,i,function(t){return e[t]}.bind(null,i));return r},n.n=function(e){var t=e&&e.__esModule?function(){return e.default}:function(){return e};return n.d(t,"a",t),t},n.o=function(e,t){return Object.prototype.hasOwnProperty.call(e,t)},n.p="",n(n.s=6)}([function(e,t,n){"use strict";e.exports=n(4)},function(e,t){e.exports=function(e){return e.webpackPolyfill||(e.deprecate=function(){},e.paths=[],e.children||(e.children=[]),Object.defineProperty(e,"loaded",{enumerable:!0,get:function(){return e.l}}),Object.defineProperty(e,"id",{enumerable:!0,get:function(){return e.i}}),e.webpackPolyfill=1),e}},function(e,t){e.exports=require("react")},function(e,t){e.exports=require("styled-components")},function(e,t,n){"use strict";Object.defineProperty(t,"__esModule",{value:!0});var r=function(e){return e&&"object"==typeof e&&"default"in e?e.default:e}(n(2)),i=function(e,t){if(!(e instanceof t))throw new TypeError("Cannot call a class as a function")},o=function(e,t){if(!e)throw new ReferenceError("this hasn't been initialised - super() hasn't been called");return!t||"object"!=typeof t&&"function"!=typeof t?e:t},s=function(e){function t(){return i(this,t),o(this,e.apply(this,arguments))}return function(e,t){if("function"!=typeof t&&null!==t)throw new TypeError("Super expression must either be null or a function, not "+typeof t);e.prototype=Object.create(t&&t.prototype,{constructor:{value:e,enumerable:!1,writable:!0,configurable:!0}}),t&&(Object.setPrototypeOf?Object.setPrototypeOf(e,t):e.__proto__=t)}(t,e),t.prototype.render=function(){return r.Children.only(this.props.children)},t}(r.Component);t.AppContainer=s,t.hot=function(){return function(e){return e}},t.areComponentsEqual=function(e,t){return e===t},t.setConfig=function(){},t.cold=function(e){return e}},,function(e,t,n){"use strict";(function(e){Object.defineProperty(t,"__esModule",{value:!0});var r=n(3),i=function(e){return e&&e.__esModule?e:{default:e}}(r),o=n(7);!function(){var t=n(0).enterModule;t&&t(e)}();var s="position: absolute; top: 0; right: 0; bottom: 0; left: 0;",u=i.default.div.withConfig({displayName:"Div___default"})(["display:",";"," "," "," "," "," "," "," "," "," "," "," "," "," "," "," "," "," "," "," "," "," "," "," "," "," "," "," "," ",""],function(e){return e.inline?e.block?"inline-block":e.grid?"inline-grid":!0===e.flex?"inline-flex":"inline":!0===e.flex?"flex":e.block?"block":e.grid?"grid":"flex"},function(e){if(!e.block||!e.inline||e.flex){var t="";return e.row&&(t+="flex-direction: row;"),e.rowReverse&&(t+="flex-direction: row-reverse;"),e.column&&(t+="flex-direction: column;"),e.columnReverse&&(t+="flex-direction: column-reverse;"),e.justifyContent&&(t+="justify-content: "+e.justifyContent+";"),e.justifyStart&&(t+="justify-content: flex-start;"),e.justifyEnd&&(t+="justify-content: flex-end;"),e.justifyCenter&&(t+="justify-content: center;"),e.justifyBetween&&(t+="justify-content: space-between;"),e.justifyAround&&(t+="justify-content: space-around;"),e.justifyEvenly&&(t+="justify-content: space-evenly;"),e.alignItems&&(t+="align-items: "+e.alignItems+";"),e.itemsStart&&(t+="align-items: flex-start;"),e.itemsEnd&&(t+="align-items: flex-end;"),e.itemsCenter&&(t+="align-items: center;"),e.itemsBaseline&&(t+="align-items: baseline;"),e.itemsStretch&&(t+="align-items: stretch;"),e.alignContent&&(t+="align-content: "+e.alignContent+";"),e.contentStart&&(t+="align-content: flex-start;"),e.contentEnd&&(t+="align-content: flex-end;"),e.contentCenter&&(t+="align-content: center;"),e.contentBetween&&(t+="align-content: space-between;"),e.contentArouns&&(t+="align-content: space-around;"),e.contentStretch&&(t+="align-content: stretch;"),e.alignSelf&&(t+="align-self: "+e.alignSelf+";"),e.selfAuto&&(t+="align-self: auto;"),e.selfStart&&(t+="align-self: flex-start;"),e.selfEnd&&(t+="align-self: flex-end;"),e.selfCenter&&(t+="align-self: center;"),e.selfBaseline&&(t+="align-self: baseline;"),e.selfStretch&&(t+="align-self: stretch;"),t}},function(e){return e.wraps&&"flex-wrap: wrap;"},function(e){var t=e.area;return"string"==typeof t&&"grid-area: "+t+";"},function(e){var t=e.order;return(0,o.isAlphaNumeric)(t)&&"order: "+t+";"},function(e){var t=e.flexNone,n=e.flex;return t?"flex: none;":(0,o.isAlphaNumeric)(n)?(0,r.css)(["flex:",";"],n):void 0},function(e){var t=e.absolute,n=e.fixed,r=e.relative;return t?"position: absolute;":n?"position: fixed;":r?"position: relative;":void 0},o.createPosition,function(e){var t=e.width;return t&&(0,r.css)(["width:",";"],(0,o.withUnit)(t))},function(e){var t=e.height;return t&&(0,r.css)(["height:",";"],(0,o.withUnit)(t))},function(e){var t=e.minWidth;return t&&(0,r.css)(["min-width:",";"],(0,o.withUnit)(t))},function(e){var t=e.minHeight;return t&&(0,r.css)(["min-height:",";"],(0,o.withUnit)(t))},function(e){var t=e.maxWidth;return t&&(0,r.css)(["max-width:",";"],(0,o.withUnit)(t))},function(e){var t=e.maxHeight;return t&&(0,r.css)(["max-height:",";"],(0,o.withUnit)(t))},function(e){return e.fullHeight&&(0,r.css)(["min-height:100vh;"])},function(e){var t=e.margin,n=e.m;return(t||n)&&(0,r.css)(["margin:",";"],(0,o.withUnit)(t||n))},(0,o.createSpaces)("margin"),function(e){var t=e.padding,n=e.p;return(t||n)&&(0,r.css)(["padding:",";"],(0,o.withUnit)(t||n))},(0,o.createSpaces)("padding"),function(e){var t=e.z;return(0,o.isAlphaNumeric)(t)&&(0,r.css)(["z-index:",";"],t)},o.createLists,function(e){return e.layer&&s},function(e){var t=e.square;return t&&(0,r.css)(["width:",";height:",";"],(0,o.withUnit)(t),(0,o.withUnit)(t))},function(e){return e.clickable&&"cursor: pointer;"},function(e){return e.noPointerEvents&&(0,r.css)(["pointer-events:none;"])},function(e){var t=e.background;return t&&(0,r.css)(["background:",";"],t)},function(e){var t=e.backgroundImage;return t&&(0,r.css)(["background-image:url(",");"],t)},function(e){var t=e.cover,n=e.contain;return(t||n)&&(0,r.css)(["background-size:",";"],t?"cover":"contain")},function(e){return e.overlay&&(0,r.css)(["position:",';&::after{content:"";'," background:",";opacity:0.2;pointer-events:none;}"],e.absolute?"absolute":"relative",s,e.overlay)},function(e){return void 0!==e.visible&&(0,r.css)(["transition:opacity 0.3s;",""],!1===e.visible&&"opacity: 0;")});t.default=u,function(){var t=n(0).default,r=n(0).leaveModule;t&&(t.register(s,"layerStyles","/Users/robert/dev/styled-kit/src/components/Div.js"),t.register(u,"default","/Users/robert/dev/styled-kit/src/components/Div.js"),r(e))}()}).call(this,n(1)(e))},function(e,t,n){"use strict";(function(e){Object.defineProperty(t,"__esModule",{value:!0}),function(){var t=n(0).enterModule;t&&t(e)}();var r=t.isNumber=function(e){return"number"==typeof e},i=t.isString=function(e){return"string"==typeof e},o=t.isAlphaNumeric=function(e){return i(e)||r(e)},s=t.withUnit=function(e){return r(e)?e+"px":e},u=["Left","Right","Top","Bottom"],c=t.createSpaces=function(e){return function(t){var n="";return u.forEach(function(r){var i=t[""+e[0]+r];void 0!==i&&(n+=e+"-"+r.toLowerCase()+": "+s(i)+";")}),n}},a=t.createPosition=function(e){var t="";return u.forEach(function(n){var r=e[""+n.toLowerCase()];void 0!==r&&(t+=n.toLowerCase()+": "+s(r)+";")}),t},l=t.createLists=function(e){var t="";return u.forEach(function(n,r){var i=e["list"+n];if(void 0!==i){var o=!0===i?"8px":s(i);t+="> *:not(:"+(r%2?"last":"first")+"-child) { margin-"+n.toLowerCase()+": "+o+"; }"}}),t};!function(){var t=n(0).default,f=n(0).leaveModule;t&&(t.register(r,"isNumber","/Users/robert/dev/styled-kit/src/utils/index.js"),t.register(i,"isString","/Users/robert/dev/styled-kit/src/utils/index.js"),t.register(o,"isAlphaNumeric","/Users/robert/dev/styled-kit/src/utils/index.js"),t.register(s,"withUnit","/Users/robert/dev/styled-kit/src/utils/index.js"),t.register(u,"directions","/Users/robert/dev/styled-kit/src/utils/index.js"),t.register(c,"createSpaces","/Users/robert/dev/styled-kit/src/utils/index.js"),t.register(a,"createPosition","/Users/robert/dev/styled-kit/src/utils/index.js"),t.register(l,"createLists","/Users/robert/dev/styled-kit/src/utils/index.js"),f(e))}()}).call(this,n(1)(e))}]);