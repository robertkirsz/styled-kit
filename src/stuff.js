import withUnit from 'utils/withUnit'

const layerStyles = 'position: absolute; top: 0; right: 0; bottom: 0; left: 0;'

export default {
  // flex-direction
  row: 'flex-direction: row;',
  rowReverse: 'flex-direction: row-reverse;',
  column: 'flex-direction: column;',
  columnReverse: 'flex-direction: column-reverse;',
  wraps: 'flex-wrap: wrap;',
  // justify-content
  justifyStart: 'justify-content: flex-start;',
  justifyEnd: 'justify-content: flex-end;',
  justifyCenter: 'justify-content: center;',
  justifyBetween: 'justify-content: space-between;',
  justifyAround: 'justify-content: space-around;',
  justifyEvenly: 'justify-content: space-evenly;',
  // align-items
  itemsStart: 'align-items: flex-start;',
  itemsEnd: 'align-items: flex-end;',
  itemsCenter: 'align-items: center;',
  itemsBaseline: 'align-items: baseline;',
  itemsStretch: 'align-items: stretch;',
  // align-content
  contentStart: 'align-content: flex-start;',
  contentEnd: 'align-content: flex-end;',
  contentCenter: 'align-content: center;',
  contentBetween: 'align-content: space-between;',
  contentAround: 'align-content: space-around;',
  contentStretch: 'align-content: stretch;',
  // Self-positioning
  flex: value => `flex: ${value};`,
  flexNone: 'flex: none;',
  selfAuto: 'align-self: auto;',
  selfStart: 'align-self: flex-start;',
  selfEnd: 'align-self: flex-end;',
  selfCenter: 'align-self: center;',
  selfBaseline: 'align-self: baseline;',
  selfStretch: 'align-self: stretch;',
  // Size
  /**/ width: value => `width: ${withUnit(value)};`,
  /**/ height: value => `height: ${withUnit(value)};`,
  /**/ minWidth: value => `min-width: ${withUnit(value)};`,
  /**/ minHeight: value => `min-height: ${withUnit(value)};`,
  /**/ maxWidth: value => `max-width: ${withUnit(value)};`,
  /**/ maxHeight: value => `max-height: ${withUnit(value)};`,
  // Margin
  /**/ margin: value => `margin: ${withUnit(value)};`,
  /**/ marginTop: value => `margin-top: ${withUnit(value)};`,
  /**/ marginRight: value => `margin-right: ${withUnit(value)};`,
  /**/ marginBottom: value => `margin-bottom: ${withUnit(value)};`,
  /**/ marginLeft: value => `margin-left: ${withUnit(value)};`,
  mTop: value => `margin-top: ${withUnit(value)};`,
  mRight: value => `margin-right: ${withUnit(value)};`,
  mBottom: value => `margin-bottom: ${withUnit(value)};`,
  mLeft: value => `margin-left: ${withUnit(value)};`,
  // Padding
  /**/ padding: value => `padding: ${withUnit(value)};`,
  /**/ paddingTop: value => `padding-top: ${withUnit(value)};`,
  /**/ paddingRight: value => `padding-right: ${withUnit(value)};`,
  /**/ paddingBottom: value => `padding-bottom: ${withUnit(value)};`,
  /**/ paddingLeft: value => `padding-left: ${withUnit(value)};`,
  pTop: value => `padding-top: ${withUnit(value)};`,
  pRight: value => `padding-right: ${withUnit(value)};`,
  pBottom: value => `padding-bottom: ${withUnit(value)};`,
  pLeft: value => `padding-left: ${withUnit(value)};`,
  // Background
  background: value => `background: ${value};`,
  backgroundImage: value => `background-image: url(${value});`,
  cover: 'background-size: cover;',
  contain: 'background-size: contain;',
  // Border
  /**/ border: value => `border: ${value};`,
  /**/ radius: value => `border-radius: ${withUnit(value)};`,
  // Position
  relative: 'position: relative;',
  absolute: 'position: absolute;',
  fixed: 'position: fixed;',
  sticky: 'position: sticky;',
  top: value => `top: ${withUnit(value)};`,
  right: value => `right: ${withUnit(value)};`,
  bottom: value => `bottom: ${withUnit(value)};`,
  left: value => `left: ${withUnit(value)};`,
  z: value => `z-index: ${value};`,
  // Text
  /**/ color: value => `color: ${value};`,
  // Lists
  listLeft: value => `> *:not(:first-child) { margin-left: ${withUnit(value === true ? 8 : value)}; }`,
  listRight: value => `> *:not(:last-child) { margin-right: ${withUnit(value === true ? 8 : value)}; }`,
  listTop: value => `> *:not(:first-child) { margin-top: ${withUnit(value === true ? 8 : value)}; }`,
  listBottom: value => `> *:not(:last-child) { margin-bottom: ${withUnit(value === true ? 8 : value)}; }`,
  // Helpers
  layer: layerStyles,
  square: value => value !== '' && `width: ${withUnit(value)}; height: ${withUnit(value)};`,
  fullHeight: 'min-height: 100vh;',
  hide: 'display: none;',
  circle: 'border-radius: 50%;',
  clickable: 'cursor: pointer;',
  noPointerEvents: 'pointer-events: none;',
  overlay: (value = 'red', props) => `
  position: ${(props.absolute && 'absolute') || 'relative'};

  &::after {
    content: "";
    ${layerStyles}
    background: ${value};
    opacity: 0.2;
    pointer-events: none;
  }`
}
