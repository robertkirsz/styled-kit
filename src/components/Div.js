import styled, { css } from 'styled-components'
import stuff from 'stuff'
import camelToKebab from 'utils/camelToKebab'
import isDivProp from 'utils/isDivProp'

function createCss(props) {
  return (previous, current) => {
    const stuffValue = stuff[current]
    const propValue = props[current]

    if (typeof stuffValue === 'string' && propValue) return `${previous}${stuffValue}`
    if (typeof stuffValue === 'function') return `${previous}${stuffValue(propValue, props)}`
    return previous
  }
}

function doStuff(props) {
  return Object.keys(props).filter(isDivProp).reduce(createCss(props), '')
}

function doMediaQueriesStuff(props = {}) {
  const { styledKitMediaQueries = {} } = props.theme || {}
  const queryNames = Object.keys(styledKitMediaQueries)

  if (!queryNames.length) return

  const queryNameToValuesMap = queryNames.reduce((all, query) => {
    const declaration = props[query]

    if (!declaration) return all

    if (typeof declaration === 'string') return { ...all, [query]: declaration }

    if (Array.isArray(declaration)) {
      return {
        ...all,
        [query]: declaration.reduce((all, property) => {
          const stuffProperty = stuff[property]

          if (typeof stuffProperty !== 'string') return all
          return `${all}${stuffProperty}`
        }, '')
      }
    }

    return {
      ...all,
      [query]: Object.keys(declaration).reduce((all, property) => {
        const value = declaration[property]
        const stuffProperty = stuff[property]

        if (!stuffProperty) return `${all}${camelToKebab(property)}:${value};`
        return `${all}${typeof stuffProperty === 'function' ? stuffProperty(value, props) : stuffProperty}`
      }, '')
    }
  }, {})

  return queryNames.reduce(
    (all, query) =>
      !queryNameToValuesMap[query]
        ? all
        : css`
            ${all} ${styledKitMediaQueries[query]`${queryNameToValuesMap[query]}`}
          `,
    ''
  )
}

export default styled.div`
  display: ${({ inline }) => (inline ? 'inline-flex' : 'flex')};
  ${doStuff}
  ${doMediaQueriesStuff}
`
