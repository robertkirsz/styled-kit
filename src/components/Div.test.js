import React from 'react'
import { ThemeProvider } from 'styled-components'
import { render } from '@testing-library/react'

import Div, { createQueries } from './Div'
import stuff from '../stuff'

export const styledKitMediaQueries = createQueries({
  small: '(max-width: 639px)',
  medium: '(min-width: 640px)',
  large: '(min-width: 1024px)'
})

const props = Object.keys(stuff).reduce((result, prop) => {
  // TODO: Ignore shorthand helper? They duplicate CSS styles.
  if (['mTop', 'mRight', 'mBottom', 'mLeft', 'pTop', 'pRight', 'pBottom', 'pLeft'].includes(prop)) return result
  return { ...result, [prop]: typeof stuff[prop] === 'function' ? 1 : true }
}, {})

describe('<Div />', () => {
  it('Matches snapshot', () => {
    const { getByTestId } = render(
      <ThemeProvider theme={{ styledKitMediaQueries }}>
        <Div
          data-testid="test"
          {...props}
          small={props}
          medium={{
            flexDirection: 'column',
            justifyContent: 'center',
            alignItems: 'center',
            background: 'pink'
          }}
          large={`
            flex-direction: column;
            justify-content: center;
            align-items: center;
            background: gray;
            > *:not(:first-child) { margin-top: 8px; }
          `}
        />
      </ThemeProvider>
    )

    expect(getByTestId('test')).toMatchSnapshot()
  })
})
