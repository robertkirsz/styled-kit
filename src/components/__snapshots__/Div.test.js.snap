// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Root Div props Border 1`] = `
.c0 {
  display: flex;
  border: 1px solid pink;
  border-radius: 50%;
}

<div
  className="c0"
  radius="50%"
/>
`;

exports[`Root Div props Flex 1`] = `
.c0 {
  display: flex;
  flex-direction: row;
}

<div
  className="c0"
/>
`;

exports[`Root Div props Size 1`] = `
.c0 {
  display: flex;
  width: 100px;
  max-width: 200px;
  height: 100%;
  max-height: 200em;
}

<div
  className="c0"
  height="100%"
  width={100}
/>
`;

exports[`Root Div props Space 1`] = `
.c0 {
  display: flex;
  margin: 16px;
  margin-top: 8px;
  margin-right: 8px;
  margin-bottom: 8px;
  margin-left: 8px;
  padding: 16px;
  padding-top: 8px;
  padding-right: 8px;
  padding-bottom: 8px;
  padding-left: 8px;
}

<div
  className="c0"
/>
`;

exports[`Root Div props Text 1`] = `
.c0 {
  display: flex;
  color: black;
}

<div
  className="c0"
  color="black"
/>
`;
